### PR TITLE
JP-3669: Moves the CHARGELOSS Read Noise Variance Recalculation to STCAL

### LINKED ISSUE #8601: Ramp fitting runtime in RN variance calculation
_Issue [JP-3669](https://jira.stsci.edu/browse/JP-3669) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

I've been using the new C-based ramp fitting code pretty extensively lately, and the one oddity that I've noticed was from processing jw02079004003_03101_00001_nis_uncal.fits

The C code runs fine, and fast (23 seconds) but then ramp fitting spends an additional 2-3 minutes before exiting from the ramp fitting routine.  I don't recall seeing a delay this lengthy before, it's usually a second or two.

Digging into a profile, it looks like the hangup is in the compute_RN_variances step (i.e., nothing to do with the C routine, but this is the next tall pole, accounting for 82% of total runtime).

Here's the time stamps:

> 2024-06-25 10:37:18,531 - stpipe.RampFitStep - INFO - Using READNOISE reference file: /Users/dlaw/crds_cache/jwst_ops/references/jwst/niriss/jwst_niriss_readnoise_0005.fits
> 
> 2024-06-25 10:37:18,531 - stpipe.RampFitStep - INFO - Using GAIN reference file: /Users/dlaw/crds_cache/jwst_ops/references/jwst/niriss/jwst_niriss_gain_0006.fits
> 
> 2024-06-25 10:37:18,572 - stpipe.RampFitStep - INFO - Using algorithm = OLS_C
> 
> 2024-06-25 10:37:18,572 - stpipe.RampFitStep - INFO - Using weighting = optimal
> 
> 2024-06-25 10:37:42,425 - stpipe.RampFitStep - INFO - Ramp Fitting C Time: 22.279687881469727
> 
> 2024-06-25 10:40:05,451 - stpipe.RampFitStep - INFO - Results used CRDS context: jwst_1241.pmap
> 
> 2024-06-25 10:40:05,453 - stpipe.RampFitStep - INFO - Results used CRDS context: jwst_1241.pmap
> 
> 2024-06-25 10:40:05,553 - stpipe.RampFitStep - INFO - Saved model in jw02079004003_03101_00001_nis_0_rampfitstep.fits
> 
> 2024-06-25 10:40:05,815 - stpipe.RampFitStep - INFO - Saved model in jw02079004003_03101_00001_nis_1_rampfitstep.fits
> 
> 2024-06-25 10:40:05,815 - stpipe.RampFitStep - INFO - Step RampFitStep done

I'm not sure why this NIRISS example is so different that the other cases I've seen.

Testing on jw01617001001_02101_00001_nrcblong_uncal.fits (8 frames/group, 5 groups/int, 10 ints) runtime is 1 second, but the NIRISS example (4 frames/group, 20 groups/int, 6 ints) is 2.5 minutes.

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves [JP-3669](https://jira.stsci.edu/browse/JP-3669)

Closes #8601

This PR moves the CHARGELOSS read noise recalculation to STCAL ramp fitting.  This PR is in conjunction with https://github.com/spacetelescope/stcal/pull/275

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/ramp_fitting/tests/test_ramp_fit.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 from stdatamodels.jwst.datamodels import dqflags, RampModel, GainModel, ReadnoiseModel
 
-from jwst.ramp_fitting.ramp_fit_step import compute_RN_variances
+# from jwst.ramp_fitting.ramp_fit_step import compute_RN_variances
```
Comment (emolter) [inline jwst/ramp_fitting/tests/test_ramp_fit.py:9]: remove comment

File: jwst/ramp_fitting/tests/test_ramp_fit.py [hunk 2]
```diff
@@ -1006,3 +952,22 @@ def base_print(label, arr):
...
+
+
+def dbg_print_dq(model):
+    nints, ngroups, nrows, ncols = model.data.shape
+    print("Model DQ:")
+    for row in range(nrows):
+        for col in range(ncols):
+            print(f"    Pixel ({row}, {col})")
+            for integ in range(nints):
+                print(f"[{integ}] {model.groupdq[integ, :, row, col]}")
```
Comment (emolter) [inline jwst/ramp_fitting/tests/test_ramp_fit.py:973]: not sure if these debug functions should remain permanently inside the repo, but I'm not too bothered either way personally. just something for maybe Tyler or Melanie to look at/decide

**Review by emolter — COMMENTED** [review]
Thanks for your good work on this Ken!  I'm a bit confused about how this fits together with the stcal changes.  It looks like a bunch of functions are removed, but no new functions are imported from stcal.  My guess is that this is all now done inside ramp_fit in the C code?  And these changes require no new keyword arguments to be passed in?  If you could just confirm that's the case, then I think I'm happy with this.

A changelog entry is needed, and the ruff style checks are failing due to an unused import that needs to be removed.

File: jwst/ramp_fitting/tests/test_ramp_fit.py (hunk 2 above)
Comment (kmacdonald-stsci) [reply jwst/ramp_fitting/tests/test_ramp_fit.py:973]: I am fine with removing them.  For debugging and test development I keep around debugging statements that I find myself rewriting over and over, but I understand if people don't want these functions in the repo.

**Comment by drlaw1558** [comment]
@kmacdonald-stsci Gotcha; so the reason for the hangup described in JP-3669 was the after-burner stuff for NIRISS handling the fact that charge_migration had been run and the RN variance needs to be tinkered with accordingly.
If this PR moves the logic entirely to the C code though, how does it handle the case where ramp fitting is run using the 'OLS' algorithm?

**Comment by kmacdonald-stsci** [comment]
> @kmacdonald-stsci Gotcha; so the reason for the hangup described in JP-3669 was the after-burner stuff for NIRISS handling the fact that charge_migration had been run and the RN variance needs to be tinkered with accordingly. If this PR moves the logic entirely to the C code though, how does it handle the case where ramp fitting is run using the 'OLS' algorithm?

It is my understanding that the "OLS" python code will be removed relatively soon.  Before this is merged, I will make sure any changes made will reflect what is wanted in the current state of the repo.  You are correct that if this gets merged as is and the "OLS" algorithm gets selected, and that selection still calls the legacy python code, then the read noise variance recalculation based on the CHARGELOSS flagging will not be done.

**Pull request closed without merging**